src/env.py

```diff
@@ -1,49 +1,615 @@
+
+    with open(key_path, "wb") as f:
+        f.write(key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption()))
+    try:
+        run(["openssl", "req", "-new", "-days", "365",
+             "-nodes", "-key", key_path, "-out",
+             csr_path, "-subj", f"/CN={username}"])
+    except CalledProcessError:
+        env_logger.error(f"Error while generating CSR for user {username}")
+        raise
+    try:
+        run(["ipa", "cert-request", csr_path, "--principal",
+             username, "--certificate-out", cert_path])
+    except CalledProcessError:
+        env_logger.error(f"Error while requesting the certificate for user "
+                         f"{username} from IPA server")
+        raise
+
+    env_logger.debug(f"User {username} is updated on IPA server. "
+                     f"Cert and key stored into {user_dir}")
+
+
+def setup_ipa_server_():
+    run(["bash", SETUP_IPA_SERVER])
+
+
+def general_setup(install_missing: bool = True):
+    """Executes script for general setup of the system. General setup includes
+    check for presense of required packages. Once this function is called,
+    READY environment variable is added to .env file and set to 1. When READY
+    is 1, script is not executed again, even if this function is called again.
+    Args:
+        install_missing: specifies if missing packages need to be automatically
+                         installed.
+    """
+
+    if read_env("READY", cast=int, default=0) != 1:
+        check_semodule()
+        packages = ["softhsm", "sssd-tools", "httpd", "sssd", "sshpass"]
+        try:
+            with open('/etc/redhat-release', "r") as f:
+                if "Red Hat Enterprise Linux release 9" not in f.read():
+                    run("dnf module enable -y idm:DL1")
+                    run("dnf install @idm:DL1 -y")
+                    env_logger.debug("idm:DL1 module is installed")
+                    
+                    run("dnf -y copr enable jjelen/vsmartcard")
+                    env_logger.debug("Copr repo for virt_cacard is enabled")
+                    
+                    run("dnf install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm -y")
+                    env_logger.debug("EPEL repository is installed")
+            run("dnf install virt_cacard vpcd -y")
+            env_logger.debug("virt_cacard is installed")
+            env_logger.debug("VPCD is installed")
+            for pkg in packages:
+                out = run(["rpm", "-qa", pkg])
+
+                if pkg not in out.stdout:
+                    if install_missing:
+                        env_logger.warning(f"Package {pkg} is not installed on the system. Installing...")
+                        run(f"dnf install {pkg} -y")
+                        env_logger.debug(f"Package {pkg} is installed")
```

Comment:
Isn't this basically the same thing as in the shell script?

Reply:
it is. I've just replaced the bash script with this. The reason for this is to make the logging of the library more transparent and make it more standard through the library. Also, in python it looks much cleaner that in bash :)  (at least for me)